File: src/use-cases/game/SelectHandSize.js
/**
 * SelectHandSize Use Case
 * Handles selecting the number of cards to deal at the start of a round
 */

const logger = require('../../../logger');

class SelectHandSize {
    /**
     * @param {IPartyRepository} partyRepository - Party repository
     * @param {IUserRepository} userRepository - User repository
     */
    constructor(partyRepository, userRepository) {
        this.partyRepository = partyRepository;
        this.userRepository = userRepository;
    }

    /**
     * Execute the use case
     * @param {Object} request - Select hand size request
     * @param {string} request.userId - User ID
     * @param {string} request.partyId - Party ID
     * @param {number} request.handSize - Number of cards to deal
     * @returns {Promise<Object>} Result
     */
    async execute({ userId, partyId, handSize }) {
        try {
            // Validate input
            if (!userId || typeof userId !== 'string') {
                throw new Error('User ID is required');
            }

            if (!partyId || typeof partyId !== 'string') {
                throw new Error('Party ID is required');
            }

            if (typeof handSize !== 'number' || !Number.isInteger(handSize)) {
                throw new Error('Hand size must be an integer');
            }

            // Verify user exists
            const user = await this.userRepository.findById(userId);
            if (!user) {
                throw new Error('User not found');
            }

            // Find party
            const party = await this.partyRepository.findById(partyId);
            if (!party) {
                throw new Error('Party not found');
            }

            // Check party is playing
            if (party.status !== 'playing') {
                throw new Error('Party is not in playing state');
            }

            // Get player
            const players = await this.partyRepository.getPartyPlayers(partyId);
            const player = players.find(p => p.userId === userId);
            if (!player) {
                throw new Error('User is not in this party');
            }

            // Get current game state
            const gameState = await this.partyRepository.getGameState(partyId);
            if (!gameState) {
                throw new Error('Game state not found');
            }

            // Check current action is selectHandSize
            if (gameState.currentAction !== 'selectHandSize') {
                throw new Error('Not in hand size selection phase');
            }

            // Verify it's this player's turn
            if (gameState.currentTurn !== player.playerIndex) {
                throw new Error('Not your turn to select hand size');
            }

            // Count active players (not eliminated)
            const eliminatedPlayers = gameState.eliminatedPlayers || [];
            const activePlayerCount = players.filter(p => !eliminatedPlayers.includes(p.playerIndex)).length;

            // Validate hand size based on player count
            // Golden Score (2 players): 4-10 cards
            // Normal (3+ players): 4-7 cards
            const isGoldenScore = activePlayerCount === 2;
            const minHandSize = 4;
            const maxHandSize = isGoldenScore ? 10 : 7;

            if (handSize < minHandSize || handSize > maxHandSize) {
                throw new Error(`Hand size must be between ${minHandSize} and ${maxHandSize} for ${activePlayerCount} players`);
            }

            // Get current deck
            const deck = [...gameState.deck];

            // Deal cards to active players only
            const hands = {};
            for (let i = 0; i < players.length; i++) {
                if (eliminatedPlayers.includes(i)) {
                    // Eliminated players get no cards
                    hands[i] = [];
                } else {
                    // Active players get cards
                    hands[i] = deck.splice(0, handSize);
                }
            }

            // Flip first card from deck to discard pile
            const firstDiscardCard = deck.pop();

            // Update game state - move to play phase
            const newGameState = gameState.with({
                deck: deck,
                hands: hands,
                lastCardsPlayed: [firstDiscardCard],
                cardsPlayed: [],
                currentAction: 'play',
                lastAction: {
                    type: 'selectHandSize',
                    playerIndex: player.playerIndex,
                    handSize: handSize
                }
            });

            // Save updated game state
            await this.partyRepository.saveGameState(partyId, newGameState);

            logger.info('Hand size selected', {
                userId: userId,
                username: user.username,
                partyId: partyId,
                handSize: handSize,
                activePlayerCount: activePlayerCount,
                isGoldenScore: isGoldenScore
            });

            return {
                success: true,
                handSize: handSize,
                gameState: {
                    currentTurn: newGameState.currentTurn,
                    currentAction: newGameState.currentAction,
                    deckSize: newGameState.deck.length,
                    handSizes: Object.keys(hands).reduce((acc, idx) => {
                        acc[idx] = hands[idx].length;
                        return acc;
                    }, {})
                }
            };
        } catch (error) {
            logger.error('Select hand size error', {
                userId,
                partyId,
                handSize,
                error: error.message
            });

            throw error;
        }
    }
}

module.exports = SelectHandSize;

File: zapzap-rust/src/application/game/select_hand_size.rs
use std::sync::Arc;

use crate::domain::entities::PartyStatus;
use crate::domain::repositories::{PartyRepository, RepositoryError};
use crate::domain::services::hand_size_bounds;
use crate::domain::value_objects::{GameAction, LastAction, LAST_ACTION_SELECT_HAND_SIZE};

/// Select hand size input
pub struct SelectHandSizeInput {
    pub party_id: String,
    pub user_id: String,
    pub hand_size: u8,
}

/// Select hand size output
pub struct SelectHandSizeOutput {
    pub hand_size: u8,
}

/// Select hand size use case
pub struct SelectHandSize<P: PartyRepository> {
    party_repo: Arc<P>,
}

impl<P: PartyRepository> SelectHandSize<P> {
    pub fn new(party_repo: Arc<P>) -> Self {
        Self { party_repo }
    }

    pub async fn execute(
        &self,
        input: SelectHandSizeInput,
    ) -> Result<SelectHandSizeOutput, SelectHandSizeError> {
        // Find party
        let party = self
            .party_repo
            .find_by_id(&input.party_id)
            .await?
            .ok_or(SelectHandSizeError::PartyNotFound)?;

        // Check party is playing
        if party.status != PartyStatus::Playing {
            return Err(SelectHandSizeError::PartyNotPlaying);
        }

        // Get player index
        let player_index = self
            .party_repo
            .get_player_index(&input.party_id, &input.user_id)
            .await?
            .ok_or(SelectHandSizeError::NotInParty)?;

        // Get game state
        let mut game_state = self
            .party_repo
            .get_game_state(&input.party_id)
            .await?
            .ok_or(SelectHandSizeError::NoGameState)?;

        // Phase, then turn, then size: the order of Node's SelectHandSize
        if game_state.current_action != GameAction::SelectHandSize {
            return Err(SelectHandSizeError::WrongAction);
        }

        if game_state.current_turn != player_index {
            return Err(SelectHandSizeError::NotYourTurn);
        }

        // 4-7 cards, 4-10 in Golden Score, and no more than the deck can deal
        let (min, max) = hand_size_bounds(&game_state);
        if input.hand_size < min || input.hand_size > max {
            return Err(SelectHandSizeError::InvalidHandSize { min, max });
        }

        // Re-deal cards with the selected hand size
        // First, collect all cards back into deck
        let mut all_cards: Vec<u8> = game_state.deck.clone();
        for i in 0..game_state.player_count as usize {
            all_cards.extend(game_state.hands[i].iter().copied());
            game_state.hands[i].clear();
        }
        all_cards.extend(game_state.last_cards_played.iter().copied());
        game_state.last_cards_played.clear();
        all_cards.extend(game_state.cards_played.iter().copied());
        game_state.cards_played.clear();
        all_cards.extend(game_state.discard_pile.iter().copied());
        game_state.discard_pile.clear();

        // Shuffle the deck
        {
            use rand::seq::SliceRandom;
            let mut rng = rand::rng();
            all_cards.shuffle(&mut rng);
        }

        // Deal cards to active players with the selected hand size
        for player in 0..game_state.player_count {
            if !game_state.is_eliminated(player) {
                for _ in 0..input.hand_size {
                    if let Some(card) = all_cards.pop() {
                        game_state.hands[player as usize].push(card);
                    }
                }
            }
        }

        // Put remaining cards back in deck
        game_state.deck = all_cards;

        // Flip first card from deck to played pile (so first player has a choice)
        if let Some(first_card) = game_state.deck.pop() {
            game_state.last_cards_played.push(first_card);
        }

        // Update action to Play
        game_state.current_action = GameAction::Play;
        game_state.last_action = LastAction {
            action_type: LAST_ACTION_SELECT_HAND_SIZE,
            player_index,
            hand_size: input.hand_size,
            ..LastAction::default()
        };

        // Save game state
        self.party_repo
            .save_game_state(&input.party_id, &game_state)
            .await?;

        // Update round
        if let Some(mut round) = self.party_repo.get_current_round(&input.party_id).await? {
            round.current_action = "play".to_string();
            self.party_repo.save_round(&round).await?;
        }

        Ok(SelectHandSizeOutput {
            hand_size: input.hand_size,
        })
    }
}

#[derive(Debug, thiserror::Error)]
pub enum SelectHandSizeError {
    #[error("Hand size must be between {min} and {max}")]
    InvalidHandSize { min: u8, max: u8 },
    #[error("Party not found")]
    PartyNotFound,
    #[error("Party is not playing")]
    PartyNotPlaying,
    #[error("Not in party")]
    NotInParty,
    #[error("No game state")]
    NoGameState,
    #[error("Not your turn")]
    NotYourTurn,
    #[error("Wrong action phase")]
    WrongAction,
    #[error("Repository error: {0}")]
    Repository(#[from] RepositoryError),
}
